### ISSUE TITLE
RSpec matchers

### ISSUE THREAD

**Issue by fabn** [aasm/aasm#147]
Are you interested in bundling some RSpec matchers in your gem?

Currently I'm using a simple matcher which only allows to check for permissible states

```ruby
describe Order do
  subject { FactoryGirl.create(:ready_order) }
  it { should be_allowed_to_go_in :created, :sent }
  it { should_not be_allowed_to_go_in :delivered }
end
```

Here's the matcher:

```ruby
  # AASM state matcher
  matcher :be_allowed_to_go_in do |*new_states|

    match do |actual|
      @exclusively ?
          new_states.to_set.subset?(permitted_states.to_set) :
          new_states.to_set.difference(permitted_states).empty?
    end

    match_when_negated do |actual|
      (new_states & permitted_states).empty?
    end

    failure_message do
      "expected #{actual.inspect} to be allowed to go in all of #{new_states}" <<
          "#{' exclusively' if @exclusively}, permissible states was #{permitted_states}"
    end

    failure_message_when_negated do
      "expected #{actual.inspect} not to be allowed to go in any of #{new_states}, permissible states was #{permitted_states}"
    end

    chain :exclusively do
      @exclusively = true
      self
    end

    def permitted_states
      @permitted_states ||= actual.aasm.states(permitted: true).map(&:name)
    end

  end

  # AASM simple event matcher
  # @example
  #   it { should allow_event :compose }
  matcher :allow_event do |event|

    match do |actual|
      actual.send("may_#{event}?")
    end

  end
```

**Comment by supertinou** [comment]
:+1:

**Comment by McRip** [comment]
:+1:

**Comment by alto** [comment]
Now I have time for new features again. Having rspec matchers or minitest assertions available for _AASM_ would be a great feature. 

Should that be part of a core _AASM_ or more an extra extension gem (e.g. `aasm_rspec` and `aasm_minitest`). What do you think?

**Comment by McRip** [comment]
In my opinion it should be a core feature. All the other gems i know that provide rspec matchers include these in their core.

**Comment by alto** [comment]
On the other side, it's increasing the gem size and is not really a core feature of a state machine. In my opinion it's an add-on.

Any other opinions out there?

**Comment by fabn** [comment]
@alto it would increase gem disk usage (by a few kilobytes) but not the memory usage because they won't be required by library code. They will need to be manually required by user in `spec_helper` file.

Lot of gems already use this approach, see for instance [carrierwave](https://github.com/carrierwaveuploader/carrierwave#testing-with-carrierwave), they provide a [single file](https://github.com/carrierwaveuploader/carrierwave/blob/master/lib/carrierwave/test/matchers.rb) to be required if used.

Having a separate gem only to provide a couple of files would be pointless in my opinion.

**Comment by fabn** [comment]
I've updated the matcher above with the version I'm currently using.

**Comment by alto** [comment]
Would these matchers be a start?

```ruby
job = Job.new # single state machine per class
job.should have_state :sleeping
job.should allow_transition_to :running
job.should allow_event :run
job.should allow_event :run!
job.should_not allow_event :clean
job.should_not allow_event :clean!

# and
simple = SimpleMultipleExample.new # multiple state machines per class
simple.should have_state :standing, :on => :move
simple.should allow_transition_to :walking, :on => :move
simple.should allow_event :run
simple.should allow_event :run!
simple.should have_state :sleeping, :on => :work
simple.should allow_transition_to :processing, :on => :work
simple.should allow_event :start
simple.should allow_event :start!
simple.should_not allow_event :stop
simple.should_not allow_event :stop!
```

**Comment by kiela** [comment]
@alto Looks good to me. Do you have already matchers for that or are you planning to write them?

**Comment by alto** [comment]
@kiela I'm planning to write them, based on the code snippets in this ticket. Do you want to help?

**Comment by alto** [comment]
Will be released with #279 . Feel free to have a look and try it out.

**Comment by alto** [comment]
Released version `4.5.0` including the new RSpec matchers.
